Flow VM Bridge FLIP (#233)
onflow/flips: application/20231222-evm-vm-bridge.md

bridging protocols common in Web 3 in order to ensure onchain transparency. A generalized solution which addresses
scalability and security concerns arising from arbitrary token bridging can better ensure the security of users on the
platform.

## User Benefit

An efficient, easy to use bridge which modifies state across VMs simultaneously in a single transaction reduces
complexity for builders and improves end-user experience for the essential economic activity of moving tokens across
VMs. The abstraction afforded to users by the bridge means that multiple EVM transaction steps can be bundled together
using Cadence to securely realize bridging as a single transaction. The availability of a secure and proven platform
capability for cross-VM bridging can ensure a consistent, higher security bar than leaving developers to implement
project specific token bridges. It also significantly reduces the effort required of developers wishing to build
applications which need to bridge tokens between VMs for an optimal user experience.

## Bridge Specification

Specification outline of the Flow VM bridge for bi-directional flow of FT and NFTs between VM states. While the spec

> references NFTs, the VM bridge will treat both FTs and NFTs alike, with the caveat that NFTs have more complexity due to
> metadata which needs to be made to work for both VMs.

Author: ```suggestion
references NFTs, the VM bridge will treat both FTs and NFTs alike, with the caveat that NFTs have more complexity due to
their more detailed metadata which needs to be made to work for both VMs.
```

### Cadence to EVM

> Breakdown of the flow for a user bridging a token across VMs from Cadence to EVM. 

Author: ```suggestion
Breakdown of the steps required to bridge a token across VMs from Cadence to EVM. 
```

#### VM bridge contract functionality

* Configure EVM-defining contract if needed
* Maintain Flow <-> EVM contract relationships
* Provide utility methods for information lookup about contracts for either state space

#### Prerequisites to user actions

1. Confirm counterparty contract deployed state, else deploy
2. Setup and configure NFT metadata
3. Safety/integrity checks as needed

#### Cadence transaction: bridge a Cadence NFT to CadenceOwnedAccount

1. Ensure prerequisites
2. Store NFT into VM bridge Cadence contract escrow
3. EVM transaction: mint corresponding NFT in EVM contract

### EVM to Cadence

Sequential breakdown of the flow for a user bridging a token from EVM to Flow. The high level paths below are broken
down into their respective forks to help with understanding.

#### VM bridge contract functionality

* Initialize Cadence NFT contract
* Maintain Flow <-> EVM contract relationships
* Provide utility methods for information lookup about contracts for either state space

> **Path 1: Cadence originated NFT returning to Cadence**

Author: That is described above in the section titled "Cadence transaction: bridge a Cadence NFT to CadenceOwnedAccount"

#### Cadence transaction: COA A bridges an EVM NFT back to Cadence

* COA A calls bridge to request bridging of EVM NFT back into Cadence
* Bridge contract calls into EVM to confirm COA A is the owner of the requested NFT. If so, process continues; otherwise
  reverts

> * Bridge contract calls into EVM contract to burn EVM NFT

Author: Maybe yeah. It's likely that by now @sisyphusSmiling has a more nuanced and clear idea of the steps involved and so hopefully he can update the FLIP accordingly when he has some time?

> * Bridge contract calls into EVM contract to burn EVM NFT

Reviewer: Will do today, revisiting this FLIP after working on the [POC PR](https://github.com/onflow/flow-evm-bridge/pull/4)

* Bridge contract calls into EVM to confirm COA B is the owner of the requested NFT. If so, process continues; otherwise
  reverts
* Bridge contract calls into EVM contract to burn EVM NFT
* Contract unlocks corresponding Cadence NFT from VM bridge contract escrow storage
* Returns NFT on call to bridge contract

**Path 2: EVM originated NFT bridging to Cadence**

#### Prerequisite steps to user actions

* VM bridge checks, else creates new NFT contract/collection if not registered
* Setup and configure NFT metadata

#### Cadence transaction: COA A wants to bridge an EVM NFT to Cadence

* Ensure prerequisites 
* Upon Cadence call to bridge, COA A must include EVM transaction approving bridge account to act on the NFT
* Bridge checks that caller is the owner of the NFT on EVM side before executing approval to Bridge account COA address
* Bridge ensures approval was successful then completes the transfer to the Bridge COA address
* Bridge validates its ownership of the NFT on EVM side after transfer, thus locking NFT to be bridged
* Bridge mints NFT from Flow bridge NFT contract and returns to caller
* Collection is configured if necessary and deposited to COA A's account

> 

Author: Good call out. @sisyphusSmiling do you mind adding that when you revisit this section? Thanks

  - Deploys new instances of FlowBridgedFT/NFT
  - Maintains knowledge of all deployed contract addresses & their Flow token identifier correspondence
  - Aids visibility of bridge contract into the EVM environment with assistive methods

- **FlowBridgedFT/NFT**
  - Define Flow-native tokens bridged from Flow to EVM
  - Point to the corresponding Flow-native contract

### Case Studies

The task of bridging FTs & NFTs between VMs can be split into four distinct cases, each with their own unique path. An
asset can either be Flow- or EVM-native, and it can either be bridged from Flow to EVM or EVM to Flow. The following
sections outline an NFT bridge path for each case.

#### Flow-Native: Flow -> EVM

*Lock in Cadence & Mint in EVM*

1. Determine if asset is Flow or EVM-native
    - Check if resource is defined in bridge-hosted NFT contract - if so, EVM-native, else Flow-native
1. Determine if BridgeNFTLocker contract has been initialized for this NFT type - if not, initialize setup
    1. Call into EVM, telling FlowBridgeFactory to deploy a new FlowBridgedNFT (ERC721) contract, passing identifying
       info about the NFT and it's source contract. Note the new EVM contract address

>     1. Derive a contract name from the NFT identifier, & deploy template-generated Cadence contract to the bridge

Reviewer 2: By template-generated Cadence contract, do you mean the locker contract? it isn't clear from what is here

Reviewer: Yes, the locker contract is generated from a template. Good callout, I'll add more detail here.

       account, passing the newly deployed EVM contract address
1. Borrow a reference to the newly deployed BridgeNFTLocker contract & passthrough bridge request
    1. Lock the NFT in the contract's Locker resource
    1. Call to FlowBridgedNFT.sol to mint to mint an NFT to the defined recipient

> 1. Locker contract calls corresponding EVM FlowBridgedNFT contract to mint NFT to the provided EVMAddress

Reviewer 2: Should there be something in here also about how the locker contract gains control of the EVM token contract? It would need to gain access to its COA, right? Not just receive the address value

Reviewer: For simplicity and traceability, all bridge contracts share a single COA. This means the locker has ownership of the corresponding ERC721 by construction since it's deployed to the account hosting all bridge contracts.

### Interfaces

> :information_source: Solidity contracts will largely be boilerplate based on common standards. Interfaces are soon
> to follow

<details>
<summary>FlowEVMBridge.cdc</summary>

```cadence
access(all) contract FlowEVMBridge {

    /// Amount of $FLOW paid to bridge
    access(all) var fee: UFix64
    /// The COA which orchestrates bridge operations in EVM
    access(self) let coa: @EVM.BridgedAccount

    /// Denotes a contract was deployed to the bridge account, could be either FlowEVMBridgeLocker or FlowEVMBridgedAsset
    access(all) event BridgeContractDeployed(type: Type, name: String, evmContractAddress: EVM.EVMAddress)

    /* --- Public NFT Handling --- */

    /// Public entrypoint to bridge NFTs from Flow to EVM - cross-account bridging supported
    ///
    /// @param token: The NFT to be bridged
    /// @param to: The NFT recipient in FlowEVM
    /// @param tollFee: The fee paid for bridging
    ///
    access(all) fun bridgeNFTToEVM(token: @{NonFungibleToken.NFT}, to: EVM.EVMAddress, tollFee: @FlowToken.Vault) {
        pre {
            tollFee.balance == self.tollAmount: "Insufficient fee paid"

>             asset.isInstance(of: Type<&{FungibleToken.Vault}>) == false: "Mixed asset types are not yet supported"

Reviewer 2: ```suggestion
            token.isInstance(of: Type<&{FungibleToken.Vault}>) == false: "Mixed asset types are not yet supported"
```

    ///
    access(all) fun bridgeNFTFromEVM(
        caller: auth(Callable) &BridgedAccount,
        calldata: [UInt8],
        id: UInt64,
        evmContractAddress: EVM.EVMAddress,
        tollFee: @FlowToken.Vault
    ): @{NonFungibleToken.NFT} {
        pre {
            tollFee.balance == self.tollAmount: "Insufficient fee paid"
            FlowEVMBridgeUtils.isEVMNFT(evmContractAddress: evmContractAddress): "Unsupported asset type"
            FlowEVMBridgeUtils.isOwnerOrApproved(ofNFT: id, owner: caller.address(), evmContractAddress: evmContractAddress):
                "Caller is not the owner of or approved for requested NFT"
        }
    }

    /* --- Public FT Handling --- */

    /// Public entrypoint to bridge NFTs from Flow to EVM - cross-account bridging supported
    ///
    /// @param vault: The FungibleToken Vault to be bridged
    /// @param to: The recipient of tokens in FlowEVM
    /// @param tollFee: The fee paid for bridging
    ///
    access(all) fun bridgeTokensToEVM(vault: @{FungibleToken.Vault}, to: EVM.EVMAddress, tollFee: @FlowToken.Vault) {
        pre {
            tollFee.balance == self.tollAmount: "Insufficient fee paid"

>             asset.isInstance(of: Type<&{NonFungibleToken.NFT}>) == false: "Mixed asset types are not yet supported"

Reviewer 2: ```suggestion
            vault.isInstance(of: Type<&{NonFungibleToken.NFT}>) == false: "Mixed asset types are not yet supported"
```

    access(account) fun borrowCOA(): &EVM.BridgedAccount
}
```

</details>

<details>
<summary>FlowEVMBridgeUtils.cdc</summary>

```cadence
/// Util contract serving all bridge contracts
access(all) contract FlowEVMBridgeUtils {

    /// Commonly used Solidity function selectors to call into EVM from bridge contracts to support call encoding
    /// e.g. ownerOf(uint256)(address), getApproved(uint256)(address), mint(address, uint256), etc.
    access(self) let functionSelectors: {String: [UInt8]}

    /// Returns the requested function selector
    access(all) fun getFunctionSelector(signature: String): [UInt8]?

    /// Identifies if an asset is Flow- or EVM-native, defined by whether a bridge contract defines it or not
    access(all) fun isFlowNative(asset: &AnyResource): Bool
    /// Identifies if an asset is Flow- or EVM-native, defined by whether a bridge-owned contract defines it or not
    access(all) fun isEVMNative(evmContractAddress: EVM.EVMAddress): Bool
    /// Identifies if an asset is ERC721 && not ERC20
    access(all) fun isEVMNFT(evmContractAddress: EVM.EVMAddress): Bool
    /// Identifies if an asset is ERC20 and not ERC721

>     access(all) fun isEVMToken(evmContractAddress: EVM.EVMAddress): Bool

Reviewer 2: Not sure if calling fungible tokens `Token` is enough to distinguish it. NFTs are Tokens also, so maybe a more descriptive name here would make sense, like ERC20 or something

Author: Yeah agreed

Reviewer: Updated each to `isERC721` and `isERC20`

(COA)](https://github.com/onflow/flips/pull/225). This Cadence resource provides access to FlowEVM, enabling encoded
calls into EVM originating from the calling COA's EVM address. In short, it's a resource that also functions as an EVM
account whose access is controlled by resource ownership instead of an offchain signature.

Since COAs are the only cross-VM objects universally available from Cadence and the identity-based access of Solidity
enables any address to receive assets, callers bridging *to* FlowEVM may bridge to any EVM address. In the opposite
direction, only COAs may initiate bridging *from* EVM since there is not yet a mechanism to initiate Cadence state
change from the EVM environment.

Hopefully, this context clarifies that bridging in either direction - Flow -> EVM & EVM -> Flow - is at all times
initiated via call to the bridge's Cadence contracts.

It's also interesting to note that any project could build an asset-specific bridge between VMs using the same
primitives and toolsets leveraged for the design below. However, the intention for this bridge is to provide public

> infrastructure to seamlessly and permissionlessly move assets between VMs without requiring the support of the asset

Reviewer: Good point, I'll add it

developers.

## Overview

The central bridge contract will act as a request router & corresponding contract registrar, additionally configuring
contracts on either side of the VM to facilitate bridge requests as they arrive. Deployed contracts are “owned” by the
bridge, but owner interactions are mediated by contract logic on either end in addition to multi-sig patterns consistent
with other core network infrastructure accounts.

> On the EVM side, a central contract factory will instantiate Solidity ERC20 & ERC721 contracts as directed by calls from

Reviewer: "deploy" would probably be a better term here

offchain system components. However, the requirement to atomically move assets between VMs prevents the inclusion of
such offchain systems as they would break atomicity and introduce the need for undesirable trust assumptions. For
example, upon bridging an NFT from Flow to EVM, an offchain listener would need to recognize a request to post metadata
to IPFS, post the metadata, and commit that URI to the defining EVM contract for the relevant NFT. We must trust that a/
the request is served, b/ metadata committed to IPFS, c/ commitment to the EVM contract succeeds and d/ that URI is
correct and contains correct metadata. Then there is the issue of IPFS storage funding.

Alternatively, if Flow projects want their metadata to be served well across VMs may take it on themselves to add
offchain IPFS metadata. This would allow the ecosystem to both maintain onchain metadata advantages as well as reach
parity with EVM platform expectations. We may then want to consider a Cadence metadata view specifically for IPFS-stored
metadata to support this use case.

Yet another alternative, it may be possible to expose an API matching that of IPFS services so that bridge-stored NFTs

> metadata could be served to IPFS clients as they would request URI material from any other provider.

Author: To confirm, this would be a fully on-chain centric solution?

Reviewer: The idea is half-baked atm. I'm imagining there would be some IPFS HTTP gateway say `https://ipfs.flow.com/ipfs/<CID>` where CID is some content-based addressing of the NFT. Requests to the gateway could be served by access/observer or other data-availability node which, upon receiving a request, would derive the information about where to locate the NFT from the CID, query for the NFT metadata and serialize it into a JSON blob conforming to ERC721 metadata standards, and returning the blob. On bridging a Flow-native NFT to FlowEVM, the IPFS URI would be derived and committed to the minted NFT in the EVM side.

This construction would mean that NFT metadata never has to leave the Cadence runtime environment, but can be served on request as if it was being retrieved from any other IPFS gateway. Of course, the engineering effort is then displaced to Cadence serialization APIs, the mechanism to construct a CID from the provided NFT content, and an IPFS-like API bolted on to some DA client-serving node.

Does that make sense?

Author: Noting the fact that we reached an internal decision to defer any development of a standalone service as described above given that it's a nice to have. For the time being we will ship with some sample Cadence scripts that will query the NFT metadata and return it in an EVM compatible format for callers

Lastly, there is also the option of creating an API similar to OpenSea's metadata API, which would allow for querying of
bridged NFT metadata from a centralized source. This would be a centralized solution, but would allow for a secondary
source of truth platforms could leverage to serve metadata to their users.

This problem and potential solutions are presented as a point of discussion and are not necessarily in scope for the
contract bridge's contract design.

#### NFT IDs

> NFT ID values are some of the most critical token metadata, indentifying each token as unique. While Flow NFTs define

Author: ```suggestion
NFT ID values are some of the most critical token metadata, identifying each token as unique. While Flow NFTs define
```
